src/client/common/client_formatter.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "client_formatter.h"
+
+#include <multipass/cli/format_utils.h>
+#include <multipass/utils.h>
+
+#include <fmt/format.h>
+#include <yaml-cpp/yaml.h>
+
+#include <sstream>
+
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+namespace mp = multipass;
+namespace mpu = multipass::utils;
+
+std::string mp::ClientFormatter::format(const mp::AliasDict& aliases) const
+{
+    std::string formatted_output;
+
+    if (preferred_format == "csv")
+        formatted_output = format_csv(aliases);
+    else if (preferred_format == "json")
+        formatted_output = format_json(aliases);
+    else if (preferred_format == "table")
+        formatted_output = format_table(aliases);
+    else if (preferred_format == "yaml")
+        formatted_output = format_yaml(aliases);
+
+    return formatted_output;
+}
+
+mp::ClientFormatter::sorted_map mp::ClientFormatter::sort_dict(const mp::AliasDict& aliases) const
+{
+    return sorted_map(aliases.cbegin(), aliases.cend());
+}
+
+std::vector<std::string> mp::ClientFormatter::escape_args(const std::vector<std::string>& args) const
+{
+    std::vector<std::string> escaped_args;
+
+    auto replace = [](const std::string& s) { return QString::fromStdString(s).replace(" ", "\\ ").toStdString(); };
+
+    std::transform(args.cbegin(), args.cend(), std::back_inserter<std::vector<std::string>>(escaped_args), replace);
+
+    return escaped_args;
+}
```

Comment:
Just escaping space is going to bite us hard. What if there's semicolons or exclamation marks in the alias?

If anything, we'd need to quote as appropriate and escape quotes themselves.

Reply:
Again, we removed arguments. For the time being, this is not anymore relevant.